app/models/concerns/hyrax/collection_nesting.rb

```diff
@@ -13,8 +13,20 @@ module CollectionNesting
       define_model_callbacks :update_index, only: :after
       after_update_index :update_nested_collection_relationship_indices
       after_destroy :update_child_nested_collection_relationship_indices
+      before_save :before_update_nested_collection_relationship_indices
+      after_save :after_update_nested_collection_relationship_indices
+
+      def before_update_nested_collection_relationship_indices
+        @during_save = true
+      end
+
+      def after_update_nested_collection_relationship_indices
+        @during_save = false
```

Comment:
Even though I suggested it, I wonder if this is entirely safe to do.  Would it be possible for the save callback chain to terminate before running the after_save callback and leave the object in a bad state such that subsequent `.update_index` calls on it wouldn't fire the nested reindexing?  I'd feel more comfortable if someone with a lot of rails experience could review this.  @jeremyf @jcoyne @no-reply

Reply:
In reviewing the documentation (http://guides.rubyonrails.org/active_record_callbacks.html#available-callbacks) and thinking through this, I'd imagine we could enter a state in which the object fails to save (e.g. Fedora went away). In this case, the instance variable wouldn't matter, as we failed to write to Fedora and there are far greater problems. I see it key that the validations have already run before we hit the `before_save` callback.

### Creating an Object
* before_validation
* after_validation
* before_save
* around_save
* before_create
* around_create
* after_create
* after_save
* after_commit/after_rollback

### Updating an Object
* before_validation
* after_validation
* before_save
* around_save
* before_update
* around_update
* after_update
* after_save
* after_commit/after_rollback

Reply:
To directly answer @cjcolvar, I think the solution is an improvement on the current state. I appreciate keeping an eye open for possible data integrity issues.

Reply:
I did quite a bit of debugging through this as part of UI testing. In addition to testing the embargoes and leases, I also I tried the persistence service, reindex_all!, and the visibility changes in all of the areas where we had previously added update_index. In each case, this worked quite well.

Creating a new collection does update_index on the collection three times. Reindexer is now called once rather than three times.

Creating a new work did update_index on the work at least six times (it was hard to keep track!) Previously this meant 6 nested reindexer calls. It now does two as part of the after_save callbacks, and two others which were apparently separate update_index callbacks not part of a save.

The persistence service is still a bit problematic. `child.save` after the member_of_collections change is needed only to force a final update_index. The save itself is unnecessary. The process that happens is:
- `child.member_of_collections << parent` adds the ldp contains member_of_collections to the child. 
- parent is then indexed, and the nesting reindexer triggers from it. (This is unnecessary, as there are no changes to parent, but that's beside the point)
- after this parent reindex, the ldp contains relationship is added to the member_of_collections object
At this point, everything is complete in Fedora. However child never reindexes at all, so the subsequent `child.save` is what forces that reindex. `hyrax/app/services/hyrax/collections/nested_collection_persistence_service.rb` could substitute update_index on the child instead of `child.save`

This still seems like an error to me, and something that needs to be discussed further.